File: sdk/python/tempmail_sdk/providers/mailtemp_cc.py
"""
MailTemp.cc 渠道 — https://mailtemp.cc
PHP POST form-urlencoded API，无需认证。
创建收件箱: POST /api.php body: action=inbox，返回 JSON 字符串格式的用户名。
获取邮件: POST /api.php body: action=fetch&inbox={username}&last_id=0。
查看邮件详情: POST /api.php body: action=view&id={id}&inbox={username}。
域名: @neocea.com
"""

import json
from typing import List

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "mailtemp-cc"
BASE = "https://mailtemp.cc"
DOMAIN = "neocea.com"

# 公共请求头
_HEADERS = {
    "Content-Type": "application/x-www-form-urlencoded",
    "Accept": "*/*",
    "Referer": f"{BASE}/",
    "Origin": BASE,
}


def generate_email(channel: str = CHANNEL) -> EmailInfo:
    """
    创建 mailtemp.cc 临时邮箱

    流程:
    1. POST /api.php body: action=inbox 获取用户名
    2. 返回值为 JSON 字符串格式（带引号），需用 json.loads() 解析
    3. 拼接完整邮箱地址: {username}@neocea.com
    4. token 存储用户名（@ 前面的部分）
    """
    resp = tm_http.post(
        f"{BASE}/api.php",
        headers=_HEADERS,
        data="action=inbox",
    )
    resp.raise_for_status()

    # 返回值为 JSON 字符串格式（如 "vindictiverate"），需去掉引号
    raw_text = resp.text.strip()
    try:
        username = json.loads(raw_text)
    except (json.JSONDecodeError, ValueError):
        # 兜底：如果不是 JSON 字符串格式，直接使用原始文本
        username = raw_text

    if not username or not isinstance(username, str):
        raise RuntimeError(f"mailtemp-cc: 返回的用户名无效: {raw_text!r}")

    address = f"{username}@{DOMAIN}"
    return EmailInfo(channel=channel, email=address, _token=username)


def get_emails(email: str, token: str) -> List[Email]:
    """
    获取 mailtemp.cc 邮件列表

    流程:
    1. POST /api.php body: action=fetch&inbox={token}&last_id=0 获取邮件列表
    2. 遍历列表，对每封邮件 POST action=view&id={id}&inbox={token} 获取详情
    3. 使用 normalize_email 标准化返回结果
    """
    if not token:
        raise ValueError("mailtemp-cc: token 不能为空")

    addr = (email or "").strip()
    if not addr:
        raise ValueError("mailtemp-cc: 邮箱地址不能为空")

    # 获取邮件列表
    resp = tm_http.post(
        f"{BASE}/api.php",
        headers=_HEADERS,
        data=f"action=fetch&inbox={token}&last_id=0",
    )
    resp.raise_for_status()

    raw_list = resp.json()
    if not isinstance(raw_list, list):
        return []

    out: List[Email] = []
    for item in raw_list:
        if not isinstance(item, dict):
            continue

        mail_id = item.get("id", "")
        if not mail_id:
            continue

        # 获取邮件详情（包含 body_html）
        detail_resp = tm_http.post(
            f"{BASE}/api.php",
            headers=_HEADERS,
            data=f"action=view&id={mail_id}&inbox={token}",
        )
        detail_resp.raise_for_status()
        detail = detail_resp.json()
        if not isinstance(detail, dict):
            detail = {}

        raw = {
            "id": str(mail_id),
            "from": item.get("sender_email", item.get("sender", "")),
            "to": addr,
            "subject": item.get("subject", detail.get("subject", "")),
            "text": "",
            "html": detail.get("body_html", ""),
            "date": item.get("received_at", detail.get("received_at", "")),
        }
        out.append(normalize_email(raw, addr))

    return out

File: sdk/ruby/lib/tempmail_sdk/providers/mailtemp_cc.rb
# frozen_string_literal: true

require "uri"

module TempmailSdk
  module Providers
    # mailtemp.cc 渠道实现（PHP POST form-urlencoded API）
    #
    # 流程：POST api.php body: action=inbox 创建邮箱（返回 JSON 字符串用户名）
    #       POST api.php body: action=fetch&inbox={username}&last_id=0 获取邮件列表
    #       POST api.php body: action=view&id={id}&inbox={username} 获取邮件详情
    # token 存储 username（@前面的部分），域名固定 neocea.com
    module MailtempCc
      CHANNEL = "mailtemp-cc"
      API_URL = "https://mailtemp.cc/api.php"

      HEADERS = {
        "Content-Type" => "application/x-www-form-urlencoded",
        "Accept" => "application/json, text/plain, */*",
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8",
        "Referer" => "https://mailtemp.cc/",
        "Origin" => "https://mailtemp.cc",
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
                        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
      }.freeze

      module_function

      # 创建 mailtemp.cc 临时邮箱
      # @return [EmailInfo]
      def generate_email
        resp = Http.post(API_URL, headers: HEADERS, body: "action=inbox", timeout: 15)
        resp.raise_for_status
        # 返回值为 JSON 字符串格式（带引号），如 "vindictiverate"
        username = resp.json
        username = username.to_s.strip
        raise "mailtemp-cc: 返回的用户名为空" if username.empty?

        EmailInfo.new(channel: CHANNEL, email: "#{username}@neocea.com", token: username)
      end

      # 获取 mailtemp.cc 邮件列表
      # @param email [String]
      # @param token [String] username
      # @return [Array<Email>]
      def get_emails(email, token)
        token = token.to_s.strip
        raise "mailtemp-cc: token 为空" if token.empty?

        form = URI.encode_www_form("action" => "fetch", "inbox" => token, "last_id" => "0")
        resp = Http.post(API_URL, headers: HEADERS, body: form, timeout: 15)
        resp.raise_for_status
        items = resp.json
        return [] unless items.is_a?(Array)
        return [] if items.empty?

        items.filter_map do |item|
          next unless item.is_a?(Hash)

          mail_id = item["id"].to_s
          unless mail_id.empty?
            detail = view_email(token, mail_id)
            if detail.is_a?(Hash)
              item["html"] = detail["body_html"] if detail["body_html"]
              item["body_html"] = detail["body_html"] if detail["body_html"]
            end
          end

          item["from"] ||= item["sender_email"] || item["sender"]
          item["date"] ||= item["received_at"]
          item["to"] = email
          Normalize.normalize_email(item, email)
        end
      end

      # 获取单封邮件详情
      # @param inbox [String] username
      # @param mail_id [String]
      # @return [Hash, nil]
      def view_email(inbox, mail_id)
        form = URI.encode_www_form("action" => "view", "id" => mail_id, "inbox" => inbox)
        resp = Http.post(API_URL, headers: HEADERS, body: form, timeout: 15)
        return nil unless resp.ok?

        resp.json
      rescue StandardError
        nil
      end
    end
  end
end
